### ISSUE TITLE
test_backward_python_udf_error is flaky

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#33348]
## 🐛 Bug

This test is flaky in `dist_autograd_test`:
https://app.circleci.com/jobs/github/pytorch/pytorch/4490663
https://app.circleci.com/jobs/github/pytorch/pytorch/4477746

(I initially thought it might be due to https://github.com/pytorch/pytorch/pull/32957, but there were failures before those changes as well)

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jjlilley

**Comment by kostmo** [comment]
Two more failures of this test were observed this week:
* https://app.circleci.com/pipelines/github/pytorch/pytorch/159555/workflows/cb81b62e-1963-4f8f-b105-b70761e9bb84/jobs/5238114/tests
* https://app.circleci.com/pipelines/github/pytorch/pytorch/159434/workflows/6f209634-e2ca-4b17-9583-024ea0e5cfa3/jobs/5236607/tests

E.g.:
```RuntimeError: Processes 0 exited with error code 10```

**Comment by rohan-varma** [comment]
@kostmo Thanks for flagging. It looks like this is a new class of error:

```
Apr 23 23:08:56   test_backward_node_failure_python_udf (__main__.DistAutogradTestWithSpawn) ... [E thread_pool.cpp:112] Exception in thread pool task: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:575] Connection closed by peer [172.17.0.2]:3744
Apr 23 23:08:56 [E container.cpp:246] Could not release Dist Autograd Context after 20 attempts.
Apr 23 23:08:56 [E container.cpp:246] Could not release Dist Autograd Context after 20 attempts.
Apr 23 23:08:56 ok (0.914s)
Apr 23 23:08:57   test_backward_python_udf_error (__main__.DistAutogradTestWithSpawn) ... ERROR:root:Caught exception: 
Apr 23 23:08:57 RuntimeError: Error on Node 0: Error on Node 1: could not compute gradients for some functions
Apr 23 23:08:57 
Apr 23 23:08:57 During handling of the above exception, another exception occurred:
Apr 23 23:08:57 
Apr 23 23:08:57 Traceback (most recent call last):
Apr 23 23:08:57   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 177, in wrapper
Apr 23 23:08:57     fn()
Apr 23 23:08:57   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/dist_utils.py", line 80, in new_test_method
Apr 23 23:08:57     return_value = old_test_method(self, *arg, **kwargs)
Apr 23 23:08:57   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/rpc/dist_autograd_test.py", line 1425, in test_backward_python_udf_error
Apr 23 23:08:57     dist_autograd.backward(context_id, [loss.sum()])
Apr 23 23:08:57   File "/opt/conda/lib/python3.6/unittest/case.py", line 217, in __exit__
Apr 23 23:08:57     expected_regex.pattern, str(exc_value)))
Apr 23 23:08:57   File "/opt/conda/lib/python3.6/unittest/case.py", line 135, in _raiseFailure
Apr 23 23:08:57     raise self.test_case.failureException(msg)
Apr 23 23:08:57 AssertionError: "Simulate error on backward pass" does not match "Error on Node 0: Error on Node 1: could not compute gradients for some functions"
Apr 23 23:08:57 exiting process with exit code: 10
Apr 23 23:08:57 Some process exited badly, terminating rest.
```

**Comment by mrshenli** [comment]
Let's see if #38512 #38364 #38366 #38376 and #38348 are sufficient to fix this.
